testsuite/tests/statmemprof/comballoc.ml

```diff
@@ -0,0 +1,35 @@
+(* TEST
+   flags = "-g"
+   compare_programs = "false" *)
+
+open Gc.Memprof
+
+let f4 n = (n,n,n,n)
+
+let[@inline never] f n =
+  (n, (n, n, f4 n))
+
+let test sampling_rate =
+  let counts = Array.make 257 0 in
+  start ~callstack_size:10
+    ~minor_alloc_callback:(fun info ->
+      counts.(info.size) <- counts.(info.size) + info.n_samples; Some info.size)
+    ~minor_dealloc_callback:(fun sz ->
+      assert (sz = 2 || sz = 3))
+    ~sampling_rate ();
+  let iter = 100_000 in
+  let arr = Array.init iter (fun i ->
+    let (_, (_, _, x)) = Sys.opaque_identity f i in
+    x) in
+  stop ();
+  ignore (Sys.opaque_identity arr);
+  for i = 0 to 256 do
+    if counts.(i) > 0 then begin
+      let total = (i + 1) * iter in
+      Printf.printf "%d: %.2f\n" i
+        (float_of_int counts.(i) /. float_of_int total)
+    end
+  done
+
+let () =
+  List.iter test [0.42; 0.01; 0.83]
```

Comment:
How robust is this test wrt. randomness?

Also, it would be great to check that the deallocation callback is indeed called for blocks of size 2 and 3.

Reply:
I've added a comment about randomness, copied here:
```
      (* counts.(i) / total is
           Binomial(total, rate) / total
         which is approx.
           Normal(total * rate, total * rate*(1-rate)) / total
         which is
           Normal(1, rate*(1-rate) / total)
         which has stddev sqrt(rate*(1-rate)/total)
         which is less than 10^-3 for the values here.
         So, an error of 0.005 (enough to make %.2f print differently)
         is a 5-sigma event, with probability less than 3*10^-7 *)
```